Source/Core/Core/HW/GBACore.cpp

```diff
@@ -0,0 +1,517 @@
+
+  m_core->setAudioBufferSize(m_core, SAMPLES);
+  blip_set_rates(m_core->getAudioChannel(m_core, 0), GBA_ARM7TDMI_FREQUENCY, SAMPLE_RATE);
+  blip_set_rates(m_core->getAudioChannel(m_core, 1), GBA_ARM7TDMI_FREQUENCY, SAMPLE_RATE);
+  g_sound_stream->GetMixer()->SetGBAInputSampleRates(m_device_number, SAMPLE_RATE);
+
+  SetAVStream();
+  SetupEvent();
+
+  std::string game_title(256, '\0');
+  m_core->getGameTitle(m_core, game_title.data());
+  m_frontend = s_frontend_factory(m_device_number, rom_path, game_title, width, height);
+
+  m_core->reset(m_core);
+  m_running = true;
+
+  if (Config::Get(Config::MAIN_GBA_THREADS) && !Movie::IsRecordingInput() &&
+      !Movie::IsPlayingInput())
+  {
+    m_idle = true;
+    m_exit_loop = false;
+    m_thread = std::make_unique<std::thread>([this] { ThreadLoop(); });
+  }
+}
+
+Core::~Core()
+{
+  if (m_thread)
+  {
+    Flush();
+    m_exit_loop = true;
+    {
+      std::lock_guard<std::mutex> lock(m_queue_mutex);
+      m_command_cv.notify_one();
+    }
+    m_thread->join();
+    m_thread.reset();
+  }
+  mCoreConfigDeinit(&m_core->config);
+  m_core->deinit(m_core);
+}
+
+bool Core::LoadBIOS(const char* bios_path)
+{
+  VFile* vf = VFileOpen(bios_path, O_RDONLY);
+  if (!vf)
+  {
+    PanicAlertFmtT("Error: GBA{0} failed to open the BIOS in {1}", m_device_number + 1, bios_path);
+    return false;
+  }
+
+  if (!m_core->loadBIOS(m_core, vf, 0))
+  {
+    PanicAlertFmtT("Error: GBA{0} failed to load the BIOS in {1}", m_device_number + 1, bios_path);
+    vf->close(vf);
+    return false;
+  }
+
+  return true;
+}
+
+static VFile* LoadROM_Archive(const char* path)
+{
+  VFile* vf{};
+  VDir* archive = VDirOpenArchive(path);
+  if (archive)
+  {
+    VFile* vf_archive =
+        VDirFindFirst(archive, [](VFile* vf_) { return mCoreIsCompatible(vf_) == mPLATFORM_GBA; });
+    if (vf_archive)
+    {
+      size_t size = static_cast<size_t>(vf_archive->size(vf_archive));
+
+      std::vector<u8> buffer(size);
+      vf_archive->seek(vf_archive, 0, SEEK_SET);
+      vf_archive->read(vf_archive, buffer.data(), size);
+      vf_archive->close(vf_archive);
+
+      vf = VFileMemChunk(buffer.data(), size);
+    }
+    archive->close(archive);
+  }
+  return vf;
+}
+
+static VFile* LoadROM_Zip(const char* path)
+{
+  VFile* vf{};
+  unzFile zip = unzOpen(path);
+  if (zip)
+  {
+    do
+    {
+      unz_file_info info{};
+      if (unzGetCurrentFileInfo(zip, &info, nullptr, 0, nullptr, 0, nullptr, 0) != UNZ_OK ||
+          !info.uncompressed_size)
+        continue;
+
+      std::vector<u8> buffer(info.uncompressed_size);
+      if (!Common::ReadFileFromZip(zip, &buffer))
+        continue;
+
+      vf = VFileMemChunk(buffer.data(), info.uncompressed_size);
+      if (mCoreIsCompatible(vf) == mPLATFORM_GBA)
+      {
+        vf->seek(vf, 0, SEEK_SET);
+        break;
+      }
+
+      vf->close(vf);
+      vf = nullptr;
+    } while (unzGoToNextFile(zip) == UNZ_OK);
+    unzClose(zip);
+  }
+  return vf;
+}
```

Comment:
mGBA has minizip integration built in already. Why did you add a second version?

Reply:
Building with LIBMGBA_ONLY will disable every dependency, except for lzma for some reason which is convenient in our case since Dolphin doesn't have 7z support
If anything I'd rather go the opposite direction and disable lzma in mGBA and have Dolphin handle it instead

Reply:
If Dolphin is already linking against minizip it would be preferable to use mGBA's existing minizip code and just point it to the right headers.

Reply:
Dolphin's minizip doesn't have the zipping functions required for mGBA's ``vfs-zip``. We could update it but that would only introduce dead code